File: app/src/main/java/vandyke/siamobile/MainActivity.java
/*
 * Copyright (c) 2017 Nicholas van Dyke
 *
 * This file is subject to the terms and conditions defined in Licensing section of the file 'README.md'
 * included in this source code package. All rights are reserved, with the exception of what is specified there.
 */

package vandyke.siamobile;

import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.content.*;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.design.widget.Snackbar;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.app.AppCompatDelegate;
import android.support.v7.widget.Toolbar;
import android.util.Base64;
import android.util.TypedValue;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.TextView;
import android.widget.Toast;
import butterknife.BindView;
import butterknife.ButterKnife;
import vandyke.siamobile.about.AboutActivity;
import vandyke.siamobile.backend.CleanupService;
import vandyke.siamobile.backend.coldstorage.ColdStorageService;
import vandyke.siamobile.backend.siad.SiadMonitorService;
import vandyke.siamobile.backend.wallet.WalletMonitorService;
import vandyke.siamobile.dialogs.DonateDialog;
import vandyke.siamobile.files.fragments.FilesFragment;
import vandyke.siamobile.help.ModesActivity;
import vandyke.siamobile.help.fragments.FragmentSetupRemote;
import vandyke.siamobile.hosting.fragments.HostingFragment;
import vandyke.siamobile.links.LinksFragment;
import vandyke.siamobile.misc.Utils;
import vandyke.siamobile.settings.GlobalPrefsListener;
import vandyke.siamobile.settings.fragments.SettingsFragment;
import vandyke.siamobile.terminal.TerminalFragment;
import vandyke.siamobile.wallet.fragments.PaperWalletFragment;
import vandyke.siamobile.wallet.fragments.WalletFragment;

import java.util.Stack;

public class MainActivity extends AppCompatActivity {

    public static int backgroundColor;
    public static int REQUEST_MODE = 2;

    private GlobalPrefsListener globalPrefsListener;

    @BindView(R.id.drawer_layout)
    public DrawerLayout drawerLayout;
    @BindView(R.id.drawer_navigation_view)
    public NavigationView navigationView;
    private ActionBarDrawerToggle drawerToggle;

    private Stack<String> titleBackstack;
    private Stack<Integer> menuItemBackstack;
    private Stack<Class> classBackstack;
    private Fragment currentlyVisibleFragment;

    public enum Theme {
        LIGHT, DARK, AMOLED, CUSTOM
    }

    public static Theme theme;

    protected void onCreate(Bundle savedInstanceState) {
        SiaMobileApplication.prefs = PreferenceManager.getDefaultSharedPreferences(this);
        globalPrefsListener = new GlobalPrefsListener(this);
        SiaMobileApplication.prefs.registerOnSharedPreferenceChangeListener(globalPrefsListener);
        switch (SiaMobileApplication.prefs.getString("theme", "light")) {
            default:
            case "light":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                setTheme(R.style.AppTheme_Light);
                theme = Theme.LIGHT;
                break;
            case "dark":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                setTheme(R.style.AppTheme_Light);
                theme = Theme.DARK;
                break;
            case "amoled":
                setTheme(R.style.AppTheme_Amoled);
                theme = Theme.AMOLED;
                break;
            case "custom":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                setTheme(R.style.AppTheme_Custom);
                theme = Theme.CUSTOM;
                break;
        }
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main_layout);
        ButterKnife.bind(this);
        if (theme == Theme.CUSTOM) {
            byte[] b = Base64.decode(SiaMobileApplication.prefs.getString("customBgBase64", "null"), Base64.DEFAULT);
            Bitmap bitmap = BitmapFactory.decodeByteArray(b, 0, b.length);
            getWindow().setBackgroundDrawable(new BitmapDrawable(getResources(), bitmap));
        }

        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (SiaMobileApplication.prefs.getBoolean("transparentBars", false)) {
            toolbar.setBackgroundColor(android.R.color.transparent);
            getWindow().setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
            toolbar.setPadding(0, getStatusBarHeight(), 0, 0);
        }

        TypedValue a = new TypedValue();
        getTheme().resolveAttribute(android.R.attr.windowBackground, a, true);
        backgroundColor = a.data;

        titleBackstack = new Stack<>();
        menuItemBackstack = new Stack<>();
        classBackstack = new Stack<>();

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);
        // set action stuff for when drawer items are selected
        NavigationView.OnNavigationItemSelectedListener drawerListener = new NavigationView.OnNavigationItemSelectedListener() {
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                drawerLayout.closeDrawers();
                int menuItemId = item.getItemId();
                switch (menuItemId) {
                    case R.id.drawer_item_files:
                        displayFragmentClass(FilesFragment.class, "Files", menuItemId);
                        return true;
                    case R.id.drawer_item_hosting:
                        displayFragmentClass(HostingFragment.class, "Hosting", menuItemId);
                        return true;
                    case R.id.drawer_item_wallet:
                        displayFragmentClass(WalletFragment.class, "Wallet", menuItemId);
                        return true;
                    case R.id.drawer_item_terminal:
                        displayFragmentClass(TerminalFragment.class, "Terminal", menuItemId);
                        return true;
                    case R.id.drawer_item_settings:
                        displayFragmentClass(SettingsFragment.class, "Settings", menuItemId);
                        return true;
                    case R.id.drawer_item_about:
                        startActivity(new Intent(MainActivity.this, AboutActivity.class));
                        return false;
                    case R.id.drawer_item_links:
                        displayFragmentClass(LinksFragment.class, "Links", menuItemId);
                        return true;
                    case R.id.drawer_item_help:
                        startActivityForResult(new Intent(MainActivity.this, ModesActivity.class), REQUEST_MODE);
                        return false;
                    case R.id.drawer_item_donate:
                        DonateDialog.createAndShow(getFragmentManager());
                        return false;
                }
                return true;
            }
        };
        navigationView.setNavigationItemSelectedListener(drawerListener);
        drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, R.string.drawer_open, R.string.drawer_close);
        drawerLayout.addDrawerListener(drawerToggle);

        startService(new Intent(this, CleanupService.class));
        if (SiaMobileApplication.prefs.getString("operationMode", "cold_storage").equals("local_full_node"))
            startService(new Intent(this, SiadMonitorService.class));
        else if (SiaMobileApplication.prefs.getString("operationMode", "cold_storage").equals("cold_storage"))
            startService(new Intent(this, ColdStorageService.class));

        startService(new Intent(this, WalletMonitorService.class));

        switch (SiaMobileApplication.prefs.getString("startupPage", "wallet")) {
            case "files":
                displayFragmentClass(FilesFragment.class, "Files", R.id.drawer_item_files);
                break;
            case "hosting":
                displayFragmentClass(HostingFragment.class, "Hosting", R.id.drawer_item_hosting);
                break;
            case "wallet":
                displayFragmentClass(WalletFragment.class, "Wallet", R.id.drawer_item_wallet);
                break;
            case "terminal":
                displayFragmentClass(TerminalFragment.class, "Terminal", R.id.drawer_item_terminal);
                break;
        }

        if (SiaMobileApplication.prefs.getBoolean("firstTime", true)) {
            startActivityForResult(new Intent(this, ModesActivity.class), REQUEST_MODE);
            startActivity(new Intent(this, AboutActivity.class));
            SiaMobileApplication.prefs.edit().putBoolean("firstTime", false).apply();
        }
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_MODE) {
            if (resultCode == ModesActivity.PAPER_WALLET) {
                displayFragmentClass(PaperWalletFragment.class, "Generated paper wallet", null);
            } else if (resultCode == ModesActivity.COLD_STORAGE) {
                SiaMobileApplication.prefs.edit().putString("operationMode", "cold_storage").apply();
                displayFragmentClass(WalletFragment.class, "Wallet", R.id.drawer_item_wallet);
//                if (currentlyVisibleFragment instanceof WalletFragment)
//                    ((WalletFragment) currentlyVisibleFragment).replaceExpandFrame(new WalletCreateFragment());
            } else if (resultCode == ModesActivity.REMOTE_FULL_NODE) {
                SiaMobileApplication.prefs.edit().putString("operationMode", "remote_full_node").apply();
                displayFragmentClass(FragmentSetupRemote.class, "Remote setup", null);
            } else if (resultCode == ModesActivity.LOCAL_FULL_NODE) {
                displayFragmentClass(WalletFragment.class, "Wallet", R.id.drawer_item_wallet);
                if (Utils.isSiadSupported()) {
                    SiaMobileApplication.prefs.edit().putString("operationMode", "local_full_node").apply();
                } else
                    Toast.makeText(this, "Sorry, but your device's CPU architecture is not supported by Sia's full node", Toast.LENGTH_LONG).show();
                displayFragmentClass(WalletFragment.class, "Wallet", R.id.drawer_item_wallet);
            }
        }
    }

    public void displayFragmentClass(Class clazz, String title, Integer menuItemId) {
        String className = clazz.getSimpleName();
        FragmentManager fragmentManager = getFragmentManager();
        Fragment fragmentToBeDisplayed = fragmentManager.findFragmentByTag(className);
        FragmentTransaction transaction = fragmentManager.beginTransaction();

        if (currentlyVisibleFragment != null) {
            if (currentlyVisibleFragment == fragmentToBeDisplayed)
                return;
            transaction.hide(currentlyVisibleFragment);
        }

        transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);

        if (currentlyVisibleFragment != null)
            transaction.hide(currentlyVisibleFragment);

        if (fragmentToBeDisplayed == null) {
            try {
                fragmentToBeDisplayed = (Fragment) clazz.newInstance();
                transaction.addToBackStack(className);
                transaction.add(R.id.fragment_frame, fragmentToBeDisplayed, className);
            } catch (InstantiationException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        } else {
            transaction.show(fragmentToBeDisplayed);
        }
        setTitle(title);
        transaction.commit();
        currentlyVisibleFragment = fragmentToBeDisplayed;
        titleBackstack.push(title);
        menuItemBackstack.push(menuItemId);
        classBackstack.push(clazz);
        if (menuItemId != null)
            navigationView.setCheckedItem(menuItemId);

    }

    public void onBackPressed() {
        if (drawerLayout.isDrawerVisible(GravityCompat.START))
            drawerLayout.closeDrawer(GravityCompat.START);
        else if (titleBackstack.size() <= 1) {
            Utils.getDialogBuilder(this)
                    .setTitle("Quit?")
                    .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialogInterface, int i) {
                            finish();
                        }
                    })
                    .setNegativeButton("No", null)
                    .show();
        } else {
            titleBackstack.pop();
            menuItemBackstack.pop();
            classBackstack.pop();
            displayFragmentClass(classBackstack.pop(), titleBackstack.pop(), menuItemBackstack.pop());
        }
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        if (drawerToggle.onOptionsItemSelected(item))
            return true;
        return super.onOptionsItemSelected(item);
    }

    public void copyTextView(View view) {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("Sia text touch copy", ((TextView) view).getText());
        clipboard.setPrimaryClip(clip);
        Utils.snackbar(view, "Copied selection to clipboard", Snackbar.LENGTH_SHORT);
    }

    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        drawerToggle.syncState();
    }

    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        drawerToggle.onConfigurationChanged(newConfig);
    }

    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.toolbar_main, menu);
        return true;
    }

    public int getStatusBarHeight() {
        int result = 0;
        int resourceId = getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            result = getResources().getDimensionPixelSize(resourceId);
        }
        return result;
    }
}

File: app/src/main/java/vandyke/siamobile/ui/MainActivity.kt
/*
 * Copyright (c) 2017 Nicholas van Dyke
 *
 * This file is subject to the terms and conditions defined in 'LICENSE.md'
 */

package vandyke.siamobile.ui

import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.app.AppCompatDelegate
import android.view.MenuItem
import android.view.WindowManager
import android.widget.TextView
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*
import vandyke.siamobile.R
import vandyke.siamobile.backend.siad.SiadService
import vandyke.siamobile.ui.about.AboutFragment
import vandyke.siamobile.ui.about.SetupRemoteActivity
import vandyke.siamobile.ui.hosting.fragments.HostingFragment
import vandyke.siamobile.ui.renter.view.RenterFragment
import vandyke.siamobile.ui.settings.ModesActivity
import vandyke.siamobile.ui.settings.Prefs
import vandyke.siamobile.ui.settings.SettingsFragment
import vandyke.siamobile.ui.terminal.TerminalFragment
import vandyke.siamobile.ui.wallet.view.PaperWalletActivity
import vandyke.siamobile.ui.wallet.view.WalletFragment
import vandyke.siamobile.util.StorageUtil

class MainActivity : AppCompatActivity() {

    private lateinit var drawerToggle: ActionBarDrawerToggle

    private var visibleFragment: BaseFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        /* appearance stuff */
        if (Prefs.darkMode)
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        else
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        setTheme(R.style.AppTheme_DayNight)

        /* pass super null so that it doesn't attempt to recreate fragments. I checked the source
         * code of the super methods, and it seems it mostly just recreates fragments from the savedInstanceState,
         * which is exactly what I don't want it to do, so this shouldn't have any weird side-effects */
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setHomeButtonEnabled(true)
        if (Prefs.transparentBars) {
            toolbar.setBackgroundColor(resources.getColor(android.R.color.transparent))
            window.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
            toolbar.setPadding(0, statusBarHeight, 0, 0)
        }

        defaultTextColor = TextView(this).currentTextColor
        /* set action stuff for when drawer items are selected */
        navigationView.setNavigationItemSelectedListener({ item ->
            drawerLayout.closeDrawers()
            when (item.itemId) {
                R.id.drawer_item_renter -> displayFragment(RenterFragment::class.java)
                R.id.drawer_item_hosting -> displayFragment(HostingFragment::class.java)
                R.id.drawer_item_wallet -> displayFragment(WalletFragment::class.java)
                R.id.drawer_item_terminal -> displayFragment(TerminalFragment::class.java)
                R.id.drawer_item_settings -> displayFragment(SettingsFragment::class.java)
                R.id.drawer_item_about -> displayFragment(AboutFragment::class.java)
            }
            return@setNavigationItemSelectedListener true
        })
        drawerToggle = ActionBarDrawerToggle(this, drawerLayout, R.string.drawer_open, R.string.drawer_close)
        drawerLayout.addDrawerListener(drawerToggle)

        if (Prefs.operationMode == "local_full_node")
            startService(Intent(this, SiadService::class.java))

        if (savedInstanceState == null) {
            when (Prefs.startupPage) {
                "renter" -> displayFragment(RenterFragment::class.java)
                "hosting" -> displayFragment(HostingFragment::class.java)
                "wallet" -> displayFragment(WalletFragment::class.java)
                "terminal" -> displayFragment(TerminalFragment::class.java)
            }
        } else {
            /* find the fragment currently visible stored in the savedInstanceState */
            visibleFragment = supportFragmentManager.findFragmentByTag(savedInstanceState.getString("visibleFragment")) as BaseFragment
            setTitleAndMenuFromVisibleFragment()
        }

        if (Prefs.firstTime) {
            startActivityForResult(Intent(this, ModesActivity::class.java), REQUEST_OPERATION_MODE)
//            startActivity(Intent(this, AboutSiaActivity::class.java))
            Prefs.firstTime = false
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        /* save the visible fragment, to be retrieved in onCreate */
        if (visibleFragment != null)
            outState?.putString("visibleFragment", visibleFragment!!.javaClass.simpleName)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_OPERATION_MODE) {
            when (resultCode) {
                ModesActivity.PAPER_WALLET -> startActivity(Intent(this, PaperWalletActivity::class.java))
                ModesActivity.COLD_STORAGE -> {
                    Prefs.operationMode = "cold_storage"
                    displayFragment(WalletFragment::class.java)
                }
                ModesActivity.REMOTE_FULL_NODE -> {
                    Prefs.operationMode = "remote_full_node"
                    startActivity(Intent(this, SetupRemoteActivity::class.java))
                }
                ModesActivity.LOCAL_FULL_NODE -> {
                    displayFragment(WalletFragment::class.java)
                    if (StorageUtil.isSiadSupported) {
                        Prefs.operationMode = "local_full_node"
                    } else
                        Toast.makeText(this, "Sorry, but your device's CPU architecture is not supported by Sia's full node", Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    private fun displayFragment(clazz: Class<*>) {
        /* return if the currently visible fragment is the same class as the one we want to display */
        if (clazz == visibleFragment?.javaClass)
            return
        val tx = supportFragmentManager.beginTransaction()
        /* check if the to-be-displayed fragment already exists */
        var newFragment = supportFragmentManager.findFragmentByTag(clazz.simpleName) as? BaseFragment
        /* if not, create an instance of it and add it to the frame */
        if (newFragment == null) {
            println("NEWFRAGMENT WAS NULL")
            newFragment = clazz.newInstance() as BaseFragment
            tx.add(R.id.fragment_frame, newFragment, clazz.simpleName)
        } else {
            tx.show(newFragment)
        }
        visibleFragment?.let { tx.hide(it) }
        tx.commit()
        visibleFragment = newFragment

        setTitleAndMenuFromVisibleFragment()
    }

    private fun setTitleAndMenuFromVisibleFragment() {
        if (visibleFragment == null)
            return
        supportActionBar!!.title = visibleFragment!!.javaClass.simpleName.replace("Fragment", "")
        navigationView.setCheckedItem(when (visibleFragment!!.javaClass) {
            RenterFragment::class.java -> R.id.drawer_item_renter
            HostingFragment::class.java -> R.id.drawer_item_hosting
            WalletFragment::class.java -> R.id.drawer_item_wallet
            TerminalFragment::class.java -> R.id.drawer_item_terminal
            SettingsFragment::class.java -> R.id.drawer_item_settings
            AboutFragment::class.java -> R.id.drawer_item_about
            else -> 0 /* not sure what this should actually be, if anything */
        })
    }

    override fun onBackPressed() {
        if (drawerLayout.isDrawerVisible(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else if (visibleFragment?.onBackPressed() != true) {
            AlertDialog.Builder(this)
                    .setTitle("Quit?")
                    .setPositiveButton("Yes") { dialogInterface, i -> finish() }
                    .setNegativeButton("No", null)
                    .show()
        }
    }

    /* below methods are for drawer stuff */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (drawerToggle.onOptionsItemSelected(item))
            return true
        return super.onOptionsItemSelected(item)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        drawerToggle.syncState()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        drawerToggle.onConfigurationChanged(newConfig)
    }

    val statusBarHeight: Int
        get() {
            var result = 0
            val resourceId = resources.getIdentifier("status_bar_height", "dimen", "android")
            if (resourceId > 0) {
                result = resources.getDimensionPixelSize(resourceId)
            }
            return result
        }

    companion object {
        var defaultTextColor: Int = 0
        var REQUEST_OPERATION_MODE = 2
    }
}
